### PR TITLE
FormCommit: Unaccessible COMMITMESSAGE

### LINKED ISSUE #6908: [NBug] Access to the path 'C:\Users\michal.kulinski\source\r-us\.g...
:warning: The sections below must be filled in and this text must be removed or the issue will be closed.

## Error Details
```
System.UnauthorizedAccessException: Access to the path 'C:\Users\michal.kulinski\source\r-us\.git\COMMITMESSAGE' is denied.
   at System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath)
   at System.IO.FileStream.Init(String path, FileMode mode, FileAccess access, Int32 rights, Boolean useRights, FileShare share, Int32 bufferSize, FileOptions options, SECURITY_ATTRIBUTES secAttrs, String msgPath, Boolean bFromProxy, Boolean useLongPath, Boolean checkHost)
   at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options, String msgPath, Boolean bFromProxy, Boolean useLongPath, Boolean checkHost)
   at System.IO.StreamWriter.CreateFile(String path, Boolean append, Boolean checkHost)
   at System.IO.StreamWriter..ctor(String path, Boolean append, Encoding encoding, Int32 bufferSize, Boolean checkHost)
   at System.IO.File.InternalWriteAllText(String path, String contents, Encoding encoding, Boolean checkHost)
   at System.IO.Abstractions.FileWrapper.WriteAllText(String path, String contents, Encoding encoding)
   at GitCommands.CommitMessageManager.set_MergeOrCommitMessage(String value)
   at GitUI.CommandsDialogs.FormCommit.OnFormClosing(FormClosingEventArgs e)
   at System.Windows.Forms.Form.CheckCloseDialog(Boolean closingOnly)
```

## Environment
- Git Extensions 3.1.1.6049
- Build 2f872105b4eb42421d9fdafd68e90a9e27d20d96
- Git 2.21.0.windows.1
- Microsoft Windows NT 10.0.17763.0
- .NET Framework 4.7.3416.0
- DPI 96dpi (no scaling)

### PR DESCRIPTION
**Pull request by gerhardol**
Fixes #6908

## Proposed changes

Popup instead of NBug if the COMMITMESSAGE cannot be accessed

### Before

NBug

### After

image

image

## Test methodology 

Test by patching code only, not reproduced the problem.

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

File: GitCommands/CommitMessageManager.cs [hunk 1]
```diff
@@ -42,6 +44,10 @@ public sealed class CommitMessageManager : ICommitMessageManager
         private Encoding _commitEncoding;
         private IFileSystem _fileSystem;
 
+        private const string _cannotReadCommitMessage = "Cannot read commit message";
```
Comment (gerhardol) [inline GitCommands/CommitMessageManager.cs]: TranslationString cannot be used in GitCommands

File: GitCommands/CommitMessageManager.cs [hunk 2]
```diff
@@ -100,7 +106,20 @@ public string MergeOrCommitMessage
                 }
 
                 var (file, exists) = GetMergeOrCommitMessagePath();
-                return exists ? _fileSystem.File.ReadAllText(file, _commitEncoding) : string.Empty;
+                string result;
+                try
+                {
+                    result = exists ? _fileSystem.File.ReadAllText(file, _commitEncoding) : string.Empty;
+                }
+                catch (UnauthorizedAccessException)
```
Comment (gerhardol) [inline GitCommands/CommitMessageManager.cs]: Could be any Exception - not relevant to crash due to this

File: GitCommands/CommitMessageManager.cs (hunk 1 above)
Comment (mstv) [reply GitCommands/CommitMessageManager.cs]: They can be moved to `Strings.cs`.

File: GitCommands/CommitMessageManager.cs (hunk 2 above)
Comment (mstv) [reply GitCommands/CommitMessageManager.cs]: Yes, I'd prefer to catch all exceptions.

Comment (gerhardol) [reply GitCommands/CommitMessageManager.cs]: updated

File: GitCommands/CommitMessageManager.cs [hunk 3]
```diff
@@ -116,7 +135,16 @@ public string MergeOrCommitMessage
...
+                    try
+                    {
+                        _fileSystem.File.WriteAllText(path, content, _commitEncoding);
+                    }
+                    catch (UnauthorizedAccessException)
+                    {
+                        MessageBox.Show(null, string.Format(_cannotAccessFile, path),
+                            _cannotSaveCommitMessage, MessageBoxButtons.OK,
+                            MessageBoxIcon.Exclamation, MessageBoxDefaultButton.Button1);
+                    }
```
Comment (RussKie) [inline GitCommands/CommitMessageManager.cs]: This look a duplicated code. I propose to move it out into an own method, so we provide a consistent experience across the module. 
...And, of course, we could test it (later as `MessageBox` isn't quite test-friendly) :)

File: GitCommands/CommitMessageManager.cs (hunk 1 above)
Comment (gerhardol) [reply GitCommands/CommitMessageManager.cs]: Which file?

File: GitCommands/CommitMessageManager.cs [hunk 4]
```diff
@@ -100,7 +106,20 @@ public string MergeOrCommitMessage
...
+                string result;
+                try
+                {
+                    result = exists ? _fileSystem.File.ReadAllText(file, _commitEncoding) : string.Empty;
+                }
+                catch (UnauthorizedAccessException)
+                {
+                    MessageBox.Show(null, string.Format(_cannotAccessFile, file),
+                        _cannotReadCommitMessage, MessageBoxButtons.OK,
+                        MessageBoxIcon.Exclamation, MessageBoxDefaultButton.Button1);
```
[the remark this answers is not in the document]
Comment (gerhardol) [reply GitCommands/CommitMessageManager.cs]: ok

File: GitCommands/CommitMessageManager.cs (hunk 3 above)
Comment (gerhardol) [reply GitCommands/CommitMessageManager.cs]: The message differs slightly

I could do with ignoring the message box completely...

File: GitCommands/CommitMessageManager.cs (hunk 1 above)
Comment (mstv) [reply GitCommands/CommitMessageManager.cs]: Open the command window (menu `View` | `Other Windows` | `Command Window`) and type "open str" ...

Comment (gerhardol) [reply GitCommands/CommitMessageManager.cs]: Of course I found a file called Strings.cs, but that is still in GitUI

Comment (mstv) [reply GitCommands/CommitMessageManager.cs]: You're right. But we have another: `ResourceManager\Strings.cs`.

**Comment by RussKie** [comment]
if we missing a dedicated `Strings.cs` in GitCommands, then we need to
create one there.
ResourceManager is for globally shared strings.

**Comment by gerhardol** [comment]
Updated

However, moving strings does not make sense, they are not to be shared and cannot be made into TranslationsStrings.
It makes more sense to remove the popup completely. The user has no option anyway, this just avoids the exception. A dismiss button in NBug would be useful (upper right corner seem to be unknown). (In this case a try-catch is needed anyway.)

I do not really want a discussion here, it is wasted effort that are better spent elsewhere and that just demotivates me. This is something that very few users will see (I cannot get rid of my file, I am living with it...)

Possible solutions:
 * Hardcoded strings as in the first commit
 * Only try-catch, no popup
 * Replace property with TrySave, TryRestore, returning also the temp COMMITMESSAGE file so the popup can include that too.
 * Move CommitMessageManager.cs to GitUI

**Comment by mstv** [comment]
> Hardcoded strings as in the first commit

Let's return to this one. Those who tinker the `.git` folder in such a strange way, must live with an English message.
But we should also display `Exception.Message` (which is localized usually).

File: GitCommands/CommitMessageManager.cs [hunk 5]
```diff
@@ -100,7 +106,20 @@ public string MergeOrCommitMessage
                 }
 
                 var (file, exists) = GetMergeOrCommitMessagePath();
-                return exists ? _fileSystem.File.ReadAllText(file, _commitEncoding) : string.Empty;
+                string result;
+                try
+                {
+                    result = exists ? _fileSystem.File.ReadAllText(file, _commitEncoding) : string.Empty;
+                }
+                catch (Exception e)
```
Comment (RussKie) [inline GitCommands/CommitMessageManager.cs]: It is a convention to call this parameter `ex`. `e` is typically used for `EventArgs`
```suggestion
                catch (Exception ex)
```

File: GitCommands/CommitMessageManager.cs [hunk 6]
```diff
@@ -116,7 +135,16 @@ public string MergeOrCommitMessage
                         return;
                     }
 
-                    _fileSystem.File.WriteAllText(path, content, _commitEncoding);
+                    try
+                    {
+                        _fileSystem.File.WriteAllText(path, content, _commitEncoding);
+                    }
+                    catch (Exception e)
```
Comment (RussKie) [inline GitCommands/CommitMessageManager.cs]:
```suggestion
                    catch (Exception ex)
```

File: GitCommands/CommitMessageManager.cs (hunk 5 above)
Comment (gerhardol) [reply GitCommands/CommitMessageManager.cs]: fixed

**Comment by gerhardol** [comment]
@msftbot merge in 1 day

**Comment by gerhardol** [comment]
Rebased and squashed, no changes

**Pull request merged**